R/watcher.R

```diff
@@ -24,20 +32,63 @@ watchout <- function(handler = new_output_handler(),
   old <- options(try.outFile = con)
   defer(options(old), frame = frame)
 
-  function(plot = TRUE, incomplete_plots = FALSE) {
-    out <- list(
-      if (plot) plot_snapshot(incomplete_plots),
-      read_con(con)
-    )
-    if (!is.null(out[[1]])) {
-      handler$graphics(out[[1]])
+  capture_plot <- function(incomplete = FALSE) {
+    # only record plots for our graphics device
+    if (!identical(dev.cur(), dev)) {
+      return()
+    }
+    # cur_devs <- dev.list()
```

Comment:
@cderv I left this code in but commented out to deliberately draw your attention to it. I'm pretty sure the logic is redundant (because if the device list has changed then that implies the current device has changed) and it doesn't cause any tests to fail. I think it was included because the logic was previously split over two functions and so it was harder to see that it was redundant. But it is a change and I wanted to double check that I didn't miss anything.

Reply:
I agree this seems redundant. 

It seems all this have been introduced over time to fix different bugs. Some have tests that are passing, and other may not have test. 

I am seeing that `dev <- dev.cur()` was previously called `evaluate_top_level_expression()` so for each parsed expression 
https://github.com/r-lib/evaluate/blob/e8bd6f5783842402e5aaa46a5fb45dbfaa4eeda0/R/eval.R#L157-L163

Now we call it once when creating `watcher` with `watchout()` call, and the  check in `capture_plot()` based on initial value. 

I tried to see if this could cause problems, but it seems not.

Reply:
Yeah, I'm reasonably certain that I got that logic right, and I added a few more tests to capture various situations.

Reply:
BTW in general if you're trying to find problems it's useful to consider the distinction between `evaluate("plot(1); lines(2)")` and `evaluate("plot(1)\nlines(2)")`